source/simulator/assembly.cc

```diff
@@ -1011,46 +1048,46 @@ namespace aspect
-                                                           :
-                                                           eta * 2.0/3.0 * (scratch.div_phi_u[i] * scratch.div_phi_u[j])
-                                                          )
-                                                        - (pressure_scaling *
-                                                           scratch.div_phi_u[i] * scratch.phi_p[j])
-                                                        // finally the term -div(u). note the negative sign to make this
-                                                        // operator adjoint to the grad(p) term
-                                                        - (pressure_scaling *
-                                                           scratch.phi_p[i] * scratch.div_phi_u[j]))
-                                                      * scratch.finite_element_values.JxW(q);
-                          }
-                  }
+              const double JxW = scratch.finite_element_values.JxW(q);
+
+              for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
+                {
+                  data.local_rhs(i) += (
+                                         (density * gravity * scratch.phi_u[i])
+                                         +
+                                         // add the term that results from the compressibility. compared
+                                         // to the manual, this term seems to have the wrong sign, but this
+                                         // is because we negate the entire equation to make sure we get
+                                         // -div(u) as the adjoint operator of grad(p) (see above where
+                                         // we assemble the matrix)
+                                         (pressure_scaling *
+                                          compressibility * density *
+                                          (scratch.velocity_values[q] * gravity) *
+                                          scratch.phi_p[i])
+                                       )
+                                       * JxW;
+
+                  if (rebuild_stokes_matrix)
+                    for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
+                      {
+                        data.local_matrix(i,j) += ( (use_tensor ?
+                                                     eta * 2.0 * (scratch.grads_phi_u[i] * stress_strain_director * scratch.grads_phi_u[j])
+                                                     :
+                                                     eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
+                                                    - (use_tensor ?
+                                                       eta_two_thirds * (scratch.div_phi_u[i] * trace(stress_strain_director * scratch.grads_phi_u[j]))
```

Comment:
Wow, how completely bizarre. I learn something new every day. Thanks for figuring this out!

With this, this part of the change is obviously completely ok.